### PR TITLE
Initial support for request's unlimited attributes. Adding operation.…

### PR DESCRIPTION
**Pull request by shai-doron**
…paths/notPaths

### CODE REVIEW

File: network-config-analyzer/MultiLayerPropertiesSet.py [hunk 1]
```diff
@@ -36,72 +37,293 @@
...
         res = RequestAttrs()
-        res.methods |= self.methods
+        res.methods = self.methods.copy()
+        res.paths = self.paths.copy()
+        res.hosts = self.hosts.copy()
         return res
 
     def __and__(self, other):
         res = RequestAttrs()
-        res.methods = self.methods & other.methods
```
Comment (adisos) [inline network-config-analyzer/MultiLayerPropertiesSet.py:66]: why not using the `&` operator for methods?

File: network-config-analyzer/MultiLayerPropertiesSet.py [hunk 2]
```diff
@@ -36,72 +37,293 @@
 """
 
 
+class UnlimitedHttpAttributes:
```
Comment (adisos) [inline network-config-analyzer/MultiLayerPropertiesSet.py:40]: Found an issue with `or` operator in some example as below (the last assertion fails).
Also, consider adding unit tests for this class separately as well.

    def test_basic_1(self):
        allow_all = UnlimitedHttpAttributes(True)
        allow_nothing = UnlimitedHttpAttributes()
        a = UnlimitedHttpAttributes()
        self.assertTrue(not a)
        print(a)
        a.values = {'a'}
        b = UnlimitedHttpAttributes()
        b.values = {'a'}
        b.negation = True
        print(a)
        print(b)
        print(a | b)
        print(a & b)
        self.assertTrue(a & b == allow_nothing)
        self.assertTrue(a | b == allow_all)

File: tests/istio_testcases/example_policies/bookinfo-demo/bookinfo-policy-test-unlimited-attrs.yaml [hunk 3]
```diff
@@ -0,0 +1,52 @@
+apiVersion: security.istio.io/v1beta1
+kind: AuthorizationPolicy
+metadata:
+  name: details-viewer-add
```
Comment (adisos) [inline tests/istio_testcases/example_policies/bookinfo-demo/bookinfo-policy-test-unlimited-attrs.yaml:4]: consider renaming this, as not all operation components here are with the 'add' semantics.

File: network-config-analyzer/MultiLayerPropertiesSet.py [hunk 4]
```diff
@@ -36,72 +37,293 @@
...
+    def methods_contained_in(self, other):
+        return self.methods.issubset(other.methods)
+
     def contained_in(self, other):
-        return set.issubset(self.methods, other.methods)
+        return self.methods_contained_in(other) and \
+               self.paths.contained_in(other.paths) and \
+               self.hosts.contained_in(other.hosts)
 
     def add_methods(self, methods):
```
Comment (adisos) [inline network-config-analyzer/MultiLayerPropertiesSet.py:282]: add comment of which methods are only for testing

**Review by adisos — CHANGES_REQUESTED** [review]
Added a few comments below.

File: network-config-analyzer/IstioPolicyYamlParser.py [hunk 5]
```diff
@@ -183,21 +185,69 @@ def parse_ports_list(self, ports_list, not_ports_list):
...
+        Parse paths and notPaths elements (within an operation component of a rule)
+        :param paths_list: list[str]  list of paths
+        :param not_paths_list: list[str]  negative list of paths
+        :return: A ConnectionSet object with allowed connections
+        :rtype: ConnectionSet
+        """
+        self.validate_paths_list(paths_list)
+        self.validate_paths_list(not_paths_list)
+        request_attributes = RequestAttrs(True)
+        request_attributes.set_paths(paths_list, not_paths_list)
```
[the remark this answers is not in the document]
Comment (shai-doron) [reply network-config-analyzer/IstioPolicyYamlParser.py:228]: fixed and tested

File: network-config-analyzer/MultiLayerPropertiesSet.py (hunk 1 above)
Comment (shai-doron) [reply network-config-analyzer/MultiLayerPropertiesSet.py:66]: changed for all operators

File: network-config-analyzer/MultiLayerPropertiesSet.py (hunk 2 above)
Comment (shai-doron) [reply network-config-analyzer/MultiLayerPropertiesSet.py:40]: good catch, thanks. will fix, do some more clean up and enhance testing of this class

File: network-config-analyzer/MultiLayerPropertiesSet.py (hunk 4 above)
Comment (shai-doron) [reply network-config-analyzer/MultiLayerPropertiesSet.py:282]: done

**Comment by shai-doron** [comment]
@adisos pls review changes post code review.
I also enhanced test `bookinfo-test-request-attrs-scheme.yaml`. Yet the last query `connectivity-istio-test-unlimited-attrs` I added does not report connectivity-map as I would expect. I suspect we do not properly `OR` overlapping attributes across various policies. I assume the problem is within class `MultiLayerPropertiesSet` and looking into it. Pls advise if you see what might go wrong.

**Pull request closed without merging**